src/test/java/com/squareup/javapoet/TypeNameTest.java

```diff
@@ -41,6 +69,31 @@
     TypeName.get(recursiveEnum.getGenericParameterTypes()[0]);
   }
 
+  @Test
+  public void innerClassInGenericType() throws Exception {
+    Method genericStringInner = getClass().getDeclaredMethod("testGenericStringInner");
+    TypeName.get(genericStringInner.getReturnType());
+    TypeName.get(genericStringInner.getGenericReturnType());
+    assertNotEquals(TypeName.get(genericStringInner.getGenericReturnType()),
+        TypeName.get(getClass().getDeclaredMethod("testGenericIntInner").getGenericReturnType()));
+  }
+
+  @Test
+  public void innerGenericInGenericType() throws Exception {
+    Method genericStringInner = getClass().getDeclaredMethod("testGenericInnerString");
+    TypeName.get(genericStringInner.getReturnType());
+    TypeName.get(genericStringInner.getGenericReturnType());
+    assertNotEquals(TypeName.get(genericStringInner.getGenericReturnType()),
+        TypeName.get(getClass().getDeclaredMethod("testGenericInnerInt").getGenericReturnType()));
+  }
+
+  @Test
+  public void innerStaticInGenericType() throws Exception {
+    Method staticInGeneric = getClass().getDeclaredMethod("testNestedNonGeneric");
+    TypeName.get(staticInGeneric.getReturnType());
+    TypeName.get(staticInGeneric.getGenericReturnType());
```

Comment:
Yah agreed, I thought the same as well, but wasn't sure what to test for (in my newly added test, the important thing is simply to make sure when the generic args are different, that we actually get different type args in the ParameterizedTypeName